spec/controllers/users/registrations_controller_spec.rb

```diff
@@ -1,29 +1,284 @@
+
+    context 'when a volunteer exists' do
+      let(:user) { build(:user, id: 9999) }
+
+      before { allow(User).to receive(:find).and_return(user) }
+
+      context 'when user can view volunteer' do
+        before { allow(user).to receive(:is_visible_to_user?).and_return(true) }
+
+        it { is_expected.to be_successful }
+
+        it { expect(subject.body).to have_content(user.name) }
+      end
+
+      context 'when a user cannot view volunteer' do
+        before { allow(user).to receive(:is_visible_to_user?).and_return(false) }
+
+        it { is_expected.to_not be_successful }
+
+        it { is_expected.to redirect_to(projects_path) }
+
+        it 'displays flash error' do
+          subject
+
+          expect(flash[:error]).to eq('Sorry, no such user.') 
+        end
+      end
+    end
+
+    context 'when volunteer does not exist' do
+      let(:user) { double(id: 9999)}
+
+      it { is_expected.to_not be_successful }
+
+      it { is_expected.to redirect_to(projects_path) }
+
+      it 'displays flash error' do
+        subject
+        
+        expect(flash[:error]).to eq('Sorry, no such user.')
+      end
     end
   end
 
   describe 'GET #new' do
     it 'tracks an event' do
-      pending 'TODO'
-      fail
+      # The line below doesn't work since the GET request sets and deletes the variable within same request:
+      # expect(session[:track_event]).to eq('Project creation started')
+      expect(controller).to receive(:track_event).with('User registration started')
+
+      get :new
+    end
+
+    it 'is successful' do
+      get :new
+
+      expect(response).to be_successful
+    end
+
+    it 'displays form' do
+      get :new
+
+      expect(response.body).to include('Sign up for an account')
     end
   end
 
   describe 'POST #create' do
-    it 'tracks an event' do
-      pending 'TODO'
-      fail
+    let(:params) { { user: { email: 'test@gmail.com', password: 'test123', password_confirmation: 'test123' } } }
+
+    context 'when email does not already exist' do
+      it 'tracks an event' do
+        # The line below doesn't work since the GET request sets and deletes the variable within same request:
+        # expect(session[:track_event]).to eq('Project creation started')
+        expect(controller).to receive(:track_event).with('User registration complete')
+
+        post :create, params: params
+      end
+
+      it 'redirects to the projects page' do
+        post :create, params: params
+
+        expect(response).to redirect_to projects_path
+      end
+
+      it 'adds user to db' do
+        expect { post :create, params: params }.to change(User, :count).by(1)
+      end
+    end
+
+    context 'when email already exists' do
+      before { create(:user, email: 'test@gmail.com') }
+      
+      it 'tracks an event' do
+        expect(controller).to receive(:track_event).with('User registration complete')
+
+        post :create, params: params
+      end
```

Comment:
I'm not sure if this is expected behaviour - do we want to track an event for every successful completed registration or every attempted completed registration? If the former, then I will change the controller code to reflect this in a separate PR

Reply:
I'm not sure but I'd assume we only want tracking on success (cc @sradu )

Reply:
this is a @jamiew question

Reply:
Addressed here: https://github.com/helpwithcovid/covid-volunteers/pull/203/files#r420609540